GUACAMOLE-364: authentication and tunnel listener extension support

ceharris opened the pull request:
This PR adds support for extensions that include listeners. Listeners can implement any of the interfaces declared in the `org.apache.guacamole.ext.net.event.listener` package of the guacamole-ext module of the client. 

The fully-qualified names of classes that implement one or more of these listener interfaces are configured in the extension manifest using the `listeners` property. Listener classes are wrapped in a facade in a fashion similar to that used for authentication providers. A `ListenerService` is registered as a Guice service (in `RESTServiceModule`) and injected into the `AuthenticationService` and `TunnelRequestService`. The `ListenerService` is used to notify all bound listeners in each of the situations in which a listener is to be notified.

I made some assumptions about the desired semantics of the boolean return value of some of the listener methods in the presence of multiple listeners, since the expected behavior isn't really expressed in the contract of the listener interfaces. Similarly, I made assumptions about the effect of listener exceptions. I describe the behavior in the `ListenerService` Javadoc, although it should probably be documented as part of the extension API. My assumptions could obviously be revisited and the implementation adjusted, as well.

I used a fairly simplistic approach to loading and binding listeners. The facade that wraps an extension listener class implements all of the existing listener interfaces, and uses `instanceof` to determine whether the delegate listener actually implements a given interface. This probably works well enough for a small number of event types with a reasonably small number of listeners, but should be revisited if either of these assumptions changes.

I was a little uncertain of what should happen WRT authentication listener notification on a re-authentication. I decided that notifying listeners of the re-authentication success was what was intended. See `AuthenticationService.getAuthenticatedUser` around line 290.

necouchman commented on guacamole/src/main/java/org/apache/guacamole/tunnel/TunnelRequestService.java:
```diff
@@ -58,6 +64,74 @@
+            Credentials credentials, GuacamoleTunnel tunnel) throws GuacamoleException {
+        TunnelConnectEvent event = new TunnelConnectEvent(userContext, credentials, tunnel);
+        boolean ok = listenerService.tunnelConnected(event);
+        if (!ok) {
```
Why not just:
`if (!listenerService.tunnelConnected(event)) {`
??

In fact, you did this in the next method down...

necouchman commented on guacamole/src/main/java/org/apache/guacamole/tunnel/TunnelRequestService.java:
```diff
@@ -226,7 +300,7 @@ protected GuacamoleTunnel createConnectedTunnel(UserContext context,
      *     If an error occurs while obtaining the tunnel.
      */
-    protected GuacamoleTunnel createAssociatedTunnel(GuacamoleTunnel tunnel,
+    protected GuacamoleTunnel createAssociatedTunnel(final GuacamoleTunnel tunnel,
```
Any particular reason for changing this to final, here?  Might be a good reason, just asking...

ceharris replied:
It gets referenced in the anonymous inner `UserTunnel` type. Previously, it wasn't referenced, so this didn't need to be final.

necouchman replied:
Ah, okay, got it!

necouchman commented on guacamole-common/src/main/java/org/apache/guacamole/GuacamoleAuthenticationRejectedException.java:
```diff
@@ -0,0 +1,34 @@
+/*
```
Style issue here - should start with
`/**`
(two asterisks).

ceharris replied:
This is not a javadoc comment and should not start with two asterisks. Checking the other license headers in the project, I found none that were written with two leading asterisks.

necouchman replied:
Yes, you are correct.  Sorry about that...

necouchman commented on guacamole/src/main/java/org/apache/guacamole/extension/AuthenticationProviderFacade.java:
```diff
@@ -66,58 +65,8 @@
-        authProvider = instance;
-
+        authProvider = ProviderFactory.newInstance("authentication provider",
+            authProviderClass);
```
Not sure I understand the reason for getting rid of all of this stuff and replacing it with the single line?  I believe I see the commit log associated with this, but not sure why I understand that it's important that the debug logging be done in there vs. in here.

I'm relatively new to Java programming, so there may be a good reason for it...just asking :-).

ceharris replied:
Pulled it out so that the same logic could be reused by `ListenerFacade` rather than duplicating it.

necouchman commented on guacamole/src/main/java/org/apache/guacamole/rest/event/ListenerService.java:
```diff
@@ -0,0 +1,142 @@
+ * A service used to notify listeners registered by extensions when events of
+ * interest occur.
+ *
+ * @author Carl Harris
```
We don't do @author tags, anymore.

necouchman commented on guacamole/src/main/java/org/apache/guacamole/rest/event/ListenerService.java:
```diff
@@ -0,0 +1,142 @@
+     * are allowed to veto a tunnel connection by returning false from the
+     * listener method. Regardless of whether a particular listener rejects the
+     * tunnel connection, all listeners are notified.
+     * @param e
```
Style: add line spacing here.

necouchman commented on guacamole-common/src/main/java/org/apache/guacamole/GuacamoleAuthenticationRejectedException.java:
```diff
@@ -0,0 +1,34 @@
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
```
Extra line...

necouchman commented on guacamole-ext/src/main/java/org/apache/guacamole/net/event/listener/AuthenticationFailureListener.java:
```diff
@@ -27,8 +27,7 @@
-public interface AuthenticationFailureListener  {
-
+public interface AuthenticationFailureListener extends Listener {
     /**
```
Style: probably should leave the line spacing here.

necouchman commented on guacamole-ext/src/main/java/org/apache/guacamole/net/event/listener/Listener.java:
```diff
@@ -0,0 +1,29 @@
+ * actually implements some listener interface.
+ */
+public interface Listener {
+}
```
Style: probably should have a blank line here.

necouchman commented on guacamole/src/main/java/org/apache/guacamole/extension/Extension.java:
```diff
@@ -35,6 +35,8 @@
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipException;
 import java.util.zip.ZipFile;
+
```
Style: remove extra line.

necouchman commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerProvider.java:
```diff
@@ -0,0 +1,37 @@
+public interface ListenerProvider extends AuthenticationSuccessListener,
+        AuthenticationFailureListener, TunnelConnectListener,
+        TunnelCloseListener {
+}
```
Style: add a space here, I think....

necouchman commented on guacamole/src/main/java/org/apache/guacamole/rest/auth/AuthenticationService.java:
```diff
@@ -208,6 +218,50 @@ private AuthenticatedUser updateAuthenticatedUser(AuthenticatedUser authenticate
+    private void notifyAuthenticationSuccessListeners(
+            AuthenticatedUser authenticatedUser, GuacamoleSession session)
+            throws GuacamoleException {
+        UserContext userContext = null;
```
Style nit-pick: Probably good to add a space above this line between the header and body of the method.

necouchman commented on guacamole/src/main/java/org/apache/guacamole/rest/event/ListenerService.java:
```diff
@@ -0,0 +1,142 @@
+import org.apache.guacamole.net.event.listener.TunnelCloseListener;
+import org.apache.guacamole.net.event.listener.TunnelConnectListener;
+
+import java.util.List;
```
This should go in alphabetical order, like the rest of them.

necouchman commented on guacamole/src/main/java/org/apache/guacamole/rest/event/ListenerService.java:
```diff
@@ -0,0 +1,142 @@
+        for (AuthenticationSuccessListener listener : listeners) {
+            result = result && listener.authenticationSucceeded(e);
+        }
+        return result;
```
Any reason not to do something like:
```    
for (AuthenticationSuccessListener listener : listeners)
    if (!listener.authenticationSucceeded(e))
        return false;
return true;
```
here, and get rid of the extra boolean variable?  Same below in the tunnelConnected() and tunnelClosed() methods.

ceharris replied:
Doing so would result in some listeners not being notified when some listener determines that an authentication should be rejected. That seems inconsistent with the the concept of `listener`. For example, you might have listeners that are passing messages to external systems when an authentication success occurs and want/need to get that notification even if a some filter decides to reject the successful authentication. I documented the behavior in the preceding doc comment, and noted in the PR that the listener API doesn't really discuss the expected behavior.

ceharris commented on guacamole/src/main/java/org/apache/guacamole/rest/auth/AuthenticationService.java:
```diff
@@ -232,11 +286,17 @@ private AuthenticatedUser getAuthenticatedUser(GuacamoleSession existingSession,
 
             // Otherwise, attempt authentication as a new user
-            AuthenticatedUser authenticatedUser = AuthenticationService.this.authenticateUser(credentials);
+            AuthenticatedUser authenticatedUser = authenticateUser(credentials);
```
I suspect that my IDE noticed that the explicit `this` wasn't needed and removed it. I haven't looked at the history of this piece of code, but I suspect that it was previously refactored from a state in which defererencing `authenticateUser` in this way was necessary. I'm happy to revert this change if desired.

ceharris commented on guacamole-common/src/main/java/org/apache/guacamole/GuacamoleAuthenticationRejectedException.java:
```diff
@@ -0,0 +1,34 @@
+ * An exception thrown when a successful authentication is rejected by a
+ * AuthenticationSuccessListener in an extension.
+ */
+public class GuacamoleAuthenticationRejectedException
```
It was a judgement call. I subtyped existing exceptions to get status codes that should be appropriate for these cases, but was uncomfortable making assumptions about how the existing exception types might be interpreted in `catch` clauses. 

I'd argue that there is something fundamentally different about an API user providing an incorrect credential vs. an extension deciding that a successful authentication should be rejected. For that reason, I think a different exception type is appropriate. But I'm happy to change the code to reuse existing exceptions if that's preferred.

mike-jumper replied:
For the authentication case, there is no need for any sort of listener. This sort of veto/monitoring functionality is already present in the auth system, and is the mechanism by which things like multi-factor authentication have been implemented.

Take the Duo extension, for example:

https://github.com/apache/incubator-guacamole-client/blob/a5ebc07349bcf43c80f404836d9964ac894bed26/extensions/guacamole-auth-duo/src/main/java/org/apache/guacamole/auth/duo/DuoAuthenticationProvider.java#L82-L96

https://github.com/apache/incubator-guacamole-client/blob/a5ebc07349bcf43c80f404836d9964ac894bed26/extensions/guacamole-auth-duo/src/main/java/org/apache/guacamole/auth/duo/UserVerificationService.java#L53-L107

In the above, when the extension determines that the user needs to be verified via Duo's service, it throws an exception requesting additional credentials, implicitly halting the authentication process. This result trickles out in the response from the token REST endpoint, and the client renders a prompt requesting that they enter their code.

ceharris commented:
I addressed the various style issues and code simplification suggestions. Also reverted the unrelated change to AuthenticationService (restoring the unnecessary `AuthenticationService.this` construct that was previous present in `getAuthenticatedUser`.

Couldn't find an example in the code base of an empty interface declaration, so not sure if the proper style includes a blank line, but that can obviously be addressed if needed.

mike-jumper commented on guacamole-common/src/main/java/org/apache/guacamole/GuacamoleAuthenticationRejectedException.java:
```diff
@@ -0,0 +1,33 @@
+public class GuacamoleAuthenticationRejectedException
+    extends GuacamoleSecurityException {
+
+    public GuacamoleAuthenticationRejectedException() {
```
I haven't yet checked whether these exceptions make sense as API changes overall, but beware that the constructors of all classes need to be documented with javadoc, including exceptions.

It is also standard convention to provide the message, message/cause, and cause constructors. It would be odd if these exceptions are the only exceptions in the Guacamole API which deviate from that convention, instead providing their own default message.

See:

https://github.com/apache/incubator-guacamole-client/blob/a5ebc07349bcf43c80f404836d9964ac894bed26/guacamole-common/src/main/java/org/apache/guacamole/GuacamoleResourceClosedException.java#L30-L62

mike-jumper commented on guacamole-common/src/main/java/org/apache/guacamole/GuacamoleTunnelConnectedException.java:
```diff
@@ -0,0 +1,32 @@
+ */
+public class GuacamoleTunnelConnectedException extends GuacamoleClientException {
+
+    public GuacamoleTunnelConnectedException() {
```
I would similarly question the utility of the tunnel-specific exceptions. In my view, simply specifying that throwing any `GuacamoleException` will veto the in-progress operation (as is currently done with auth) would be cleaner, as well as in-line with established patterns.

mike-jumper commented on guacamole-ext/src/main/java/org/apache/guacamole/net/event/listener/Listener.java:
```diff
@@ -0,0 +1,28 @@
+package org.apache.guacamole.net.event.listener;
+
+/**
+ * A marker interface extended by all listener types. This interface is used
```
I would fall on the "marker interface is an anti-pattern" side of the fence. If there is no shared functional interface between these listeners, then we shouldn't be defining a contract which states that there is.

That said, if it feels wrong to not have some common ancestor, then perhaps there *should* be a shared functional interface? At the moment, the only barrier I see preventing a common function for handling events is the variation on whether a handler is `void` or `boolean`, but there may be a better approach than the way things were done in the early guac days. Doing so would simplify the way these events get fired, and would avoid all the `instanceof` checks and casting.

mike-jumper replied:
In fact ... if things move toward the style of veto used by the rest of the auth system (throw a `GuacamoleException` to abort the operation), then there definitely could be a common event handler, as there would be no need to rely on a `boolean` return value.

ceharris commented:
As opposed to having a (perhaps large) family of different listener interfaces that might be implemented, each with its own event argument types and return value semantics, I've always appreciated the simplicity (from the API user's perspective) of a listener API that has a simple functional interface with an event argument of a type that is subclassed to represent the different kinds of events that can be observed. See for example the `@Observes` event observation mechanism of CDI or the Spring Framework's `ApplicationEvent` and associated listener interface.

The simplest approach would be an interface such as the following.

```
/**
 * A listener for events that occur in handing various Guacamole requests 
 * such as authentication, tunnel connect/close, etc. Listeners are registered 
 * through the extension manifest mechanism. When an event occurs, listeners 
 * are notified in the order in which they are declared in the manifest and 
 * continues until either all listeners have been notified or with the first 
 * listener that throws a GuacamoleException or other runtime exception.
 */
public interface EventListener {

  /**
   * Notifies the recipient that an event has occurred.
   * <p>
   * Throwing an exception from an event listener can act to veto an action in 
   * progress for some event types. See the Javadoc for specific event types for
   * details.
   * 
   * @param event
   *     an object that describes the subject event
   * 
   * @throws GuacamoleException
   *     If the listener wishes to stop notification of the event to subsequent 
   *     listeners; for some event types, this acts to veto an action in progress 
   *     (e.g. an authentication).
   */
  void handleEvent(Object event) throws GuacamoleException;

}
```

One drawback to this approach is that listener implementers are responsible for identifying events of interest and acting accordingly. But features could be added later that would allow a listener to declare through some form of configuration (annotations, manifest, whatever) the subtype of events the listener is interested in hearing. Adding such features wouldn't break existing listeners that did their own event filtering, but would make it easier to implement highly targeted listeners in a manner not all that different than that which is the supported by the existing API.

ceharris commented:
This approach could even support existing listener implementations with a relatively simple adapter mechanism, if such was desirable.

mike-jumper commented:
@ceharris - I like where you're headed with that simpler, functional `EventListener` interface.

ceharris commented:
Do we care about keeping the legacy listener API?  i.e. should I make an adapter mechanism for legacy listener implementations? If we're keeping it, should we annotate the existing interfaces using `@Deprecated`?

mike-jumper commented:
> Do we care about keeping the legacy listener API? i.e. should I make an adapter mechanism for legacy listener implementations?

We care in the sense that we can't remove parts of the API (even non-functional parts) without first explicitly deprecating them. Those implementations really should have been explicitly `@Deprecated` a while ago. I'm not sure why this wasn't done, but it would need to be done now so they can be removed in some future release.

As the functionality has been broken since support for the `event-listeners` property was removed, the main concern at this point is API compatibility. There's no need to provide adapters, but I wouldn't be against it.

> If we're keeping it, should we annotate the existing interfaces using `@Deprecated`?

Definitely.

ceharris commented:
Declared and implemented new simple listener API, deprecating existing API, and including a simple adapter mechanism for legacy listener classes. The new API, by using the _veto-by-exception_ approach, eliminates the previously added exception subtypes.

mike-jumper commented:
Quickly reading through the latest changes, things look pretty solid. I like the new approach. It looks like there are some comments missing / misformatted, but I'll enumerate those via an actual review in a moment (as well as double-check that I'm not missing something obvious).

Overall, looks very nice, @ceharris.

ceharris commented:
Just re-upping this to bring it back to your attention. I see you guys have been busy with several other PRs, but wondered if there's work that can be done here to get this one ready to merge.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+    @SuppressWarnings("deprecation")
+    private static class AuthenticationSuccessListenerAdapter implements Listener {
+
+        private final AuthenticationSuccessListener delegate;
```
Missing JavaDoc.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         * Constructs a new adapter.
+         *
+         * @param delegate
+         *      the delegate listener
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         * to veto the authentication success event. All other event types are ignored.
+         *
+         * @param event
+         *     an object that describes the subject event
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         *     an object that describes the subject event
+         *
+         * @throws GuacamoleException
+         *      if thrown by the delegate listener
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+        private final AuthenticationSuccessListener delegate;
+
+        /**
+         * Constructs a new adapter.
```
Comments should add information/context which is not inherently present. "Constructs a new adapter" is what any constructor within this class would do, and certainly this constructor has some specific use dealing with its `delegate` parameter.

Consider:

    /**
     * Returns the count.
     *
     * @return
     *     The count.
     */
    public int getCount();

vs:

    /**
     * Returns the number of elements currently in this list.
     *
     * @return
     *     The number of elements in the list.
     */
    public int getCount();

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+    @SuppressWarnings("deprecation")
+    private static class AuthenticationFailureListenerAdapter implements Listener {
+
+        private final AuthenticationFailureListener delegate;
```
Missing JavaDoc.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+        private final AuthenticationFailureListener delegate;
+
+        /**
+         * Constructs a new adapter.
```
See above - "Constructs new adapter" is not sufficiently informative.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         * Constructs a new adapter.
+         *
+         * @param delegate
+         *      the delegate listener
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         * listener. All other event types are ignored.
+         *
+         * @param event
+         *     an object that describes the subject event
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         *     an object that describes the subject event
+         *
+         * @throws GuacamoleException
+         *      if thrown by the delegate listener
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+    @SuppressWarnings("deprecation")
+    private static class TunnelConnectListenerAdapter implements Listener {
+
+        private final TunnelConnectListener delegate;
```
Missing JavaDoc.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+        private final TunnelConnectListener delegate;
+
+        /**
+         * Constructs a new adapter.
```
See above - comment should be more informative.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         * Constructs a new adapter.
+         *
+         * @param delegate
+         *      the delegate listener
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         * to veto the tunnel connect event. All other event types are ignored.
+         *
+         * @param event
+         *     an object that describes the subject event
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         *     an object that describes the subject event
+         *
+         * @throws GuacamoleException
+         *      if thrown by the delegate listener
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+        private final TunnelCloseListener delegate;
+
+        /**
+         * Constructs a new adapter.
```
See above - needs to be more informative.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         * Constructs a new adapter.
+         *
+         * @param delegate
+         *      the delegate listener
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         *     an object that describes the subject event
+         *
+         * @throws GuacamoleException
+         *      if thrown by the delegate listener
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/extension/ListenerFactory.java:
```diff
@@ -0,0 +1,256 @@
+         * to veto the tunnel connect event. All other event types are ignored.
+         *
+         * @param event
+         *     an object that describes the subject event
```
Should be sentence case, plus a period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/tunnel/TunnelRequestService.java:
```diff
@@ -58,6 +62,58 @@
+     *      being created.
+     *
+     * @param credentials
+     *      Credentials that authenticate the user
```
Missing period.

mike-jumper commented on guacamole/src/main/java/org/apache/guacamole/tunnel/TunnelRequestService.java:
```diff
@@ -58,6 +62,58 @@
+     *      Credentials that authenticate the user
+     *
+     * @param tunnel
+     *      The tunnel that was connected
```
Missing period.

mike-jumper commented:
Alrighty - brace yourself. I've finished going through everything, and have a sizable review coming, but that's mainly due to copy/paste of the same thing. Will summarize once I click the "submit" button.

mike-jumper commented:
OK, summarizing:

There are some minor style issues with the documentation which should be addressed, and that makes up the majority of the noise in the review. The only problems I would consider non-minor are the occasional missing JavaDoc, and some non-informative JavaDoc.

Overall, it really boils down to the following:

* Some functions and properties are missing JavaDoc.
* Some JavaDoc is insufficiently descriptive ("Constructs a new adapter.")
* Copypasta of the same style typos (missing periods, missing blank lines between parameters, not sentence case, etc.) across similar code

I've marked everything I could find, hence the large number of requested changes, but the vast majority are the same thing repeated over and over. If you just go through and make sure (1) that your comments match the JavaDoc style being used for new code, (2) that absolutely all functions and properties are documented with JavaDoc, and (3) that your JavaDoc adds information not already present in the name of the function, I believe we're good to go.

Looking over the code itself, I do not see any issues.

On a different note, we should probably work on getting a style-checking CI job to do this... It would automate stylistic nitpicking at the front of most reviews, and allow the review itself to focus on the meat of the code. Could perhaps even allow contributors to run the same style checks against their own contributions prior to opening the PR.

ceharris commented on guacamole-ext/src/main/java/org/apache/guacamole/net/event/listener/TunnelCloseListener.java:
```diff
@@ -25,11 +25,15 @@
 
     /**
-     * Event hook which fires immediately after an existing tunnel is closed.
+     * Event hook which fires immediately before an existing tunnel is closed.
```
Should be before. Fixed at the interface level comment.

ceharris commented:
Many thanks for the painstaking code reviews!

I think this is one of the few open source projects to which I've made contributions that wasn't using checkstyle or some such to help document and enforce style expectations. And I've certainly missed having that sort of tool to do some pre-checking before subjecting someone else to having to point out such things. Definitely worth considering.
